### PR TITLE
[wmcb] Dynamically pick the kube-node package in the e2e tests

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This PR consists of 2 commits 

- Create an interface to generalize adding packages in the wmcb tests. Implement Factory pattern to retrieve package specific instances.

- The PR also ensures that tests use kube node package with the current k8s version

    **Problem**: wmcb e2e tests use kube-node packgage with hardcoded version.

   **Solution**: Use current kubernetes version to retrieve node package
   for Windows. The corresponding SHA is retrieved using changelog.

   [WINC-235](https://issues.redhat.com/browse/WINC-235)

### CODE REVIEW

File: internal/test/framework/framework.go [hunk 1]
```diff
@@ -242,6 +243,67 @@ func (f *TestFramework) getClusterVersion() error {
...
+	if err != nil {
+		return "", fmt.Errorf("error while retrieving k8s version : %s", err)
+	}
+
+	return versionInfo.GitVersion, nil
+
+}
+
+// Get KubeNodeUrl gets the url for kubernetes windows node binary.
+func (f *TestFramework) GetKubeNodeURL() (string, error) {
```
Comment (ravisantoshgudimetla) [inline internal/test/framework/framework.go]: Is this method going to be used by both wmcb and wsu? If not, this should go to `wsuFramework`.

File: internal/test/framework/framework.go [hunk 2]
```diff
@@ -242,6 +243,67 @@ func (f *TestFramework) getClusterVersion() error {
 	return nil
 }
 
+// GetK8sVersion gets the Kubernetes version in vx.x.x format
+func (f *TestFramework) GetK8sVersion() (string, error) {
+	versionInfo, err := f.OSConfigClient.Discovery().ServerVersion()
```
Comment (ravisantoshgudimetla) [inline internal/test/framework/framework.go]: You can make versionInfo a field, if you don't want make api calls to OpenShift apiserver

File: internal/test/framework/framework.go [hunk 3]
```diff
@@ -242,6 +243,67 @@ func (f *TestFramework) getClusterVersion() error {
...
+}
+
+// Get KubeNodeUrl gets the url for kubernetes windows node binary.
+func (f *TestFramework) GetKubeNodeURL() (string, error) {
+	k8sVersion, err := f.GetK8sVersion()
+
+	if err != nil {
+		return "", err
+	}
+
```
Comment (ravisantoshgudimetla) [inline internal/test/framework/framework.go]: Not sure why we're having extra blank lines

File: internal/test/framework/framework.go [hunk 4]
```diff
@@ -242,6 +243,67 @@ func (f *TestFramework) getClusterVersion() error {
...
+	if err != nil {
+		return "", fmt.Errorf("error while getting changelog file")
+	}
+
+	changeLogContent, _ := ioutil.ReadAll(changeLog.Body)
+	defer changeLog.Body.Close()
+
+	// Extract binary and corresponding SHA value
+	re := regexp.MustCompile("<tr>\n<td><a href=\"https://dl.k8s.io/" + k8sVersion +
```
Comment (ravisantoshgudimetla) [inline internal/test/framework/framework.go]: We've started moving away from re? Can't you do a strings.Split() based on the "/" and get to the value you need.

File: internal/test/wmcb/wmcb_test.go [hunk 5]
```diff
@@ -199,6 +193,23 @@ func (vm *wmcbVM) initializeTestBootstrapperFiles() error {
 		return fmt.Errorf("unable to create remote directory %s: %v", remoteDir, err)
 	}
 
+	kubeNodeURL, err := framework.GetKubeNodeURL()
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: I thinkw we can wrap getting `kubeNodeURL, KubeNodeSHA` into a single method and get the package as return value for that method.

File: internal/test/framework/framework.go (hunk 2 above)
Comment (aravindhp) [reply internal/test/framework/framework.go]: Yeah I agree. Add it to the framework struct and populate during the initialization.

File: internal/test/wmcb/wmcb_test.go (hunk 5 above)
Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go]: I agree just return `pkgInfo`

File: internal/test/framework/framework.go (hunk 1 above)
Comment (VaishnaviHire) [reply internal/test/framework/framework.go]: It is used by wmcb tests. Should I add it to `wmcbFramework` ?

Comment (ravisantoshgudimetla) [reply internal/test/framework/framework.go]: Yeah, wmcb sorry for the confusion.

File: internal/test/wmcb/wmcb_test.go [hunk 6]
```diff
@@ -677,3 +677,42 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
+func (f *wmcbFramework) getKubeNodeSHA() (string, error) {
+
+	kubeNodeChecksumFileURL := f.kubeNode.url + ".sha512"
+	// Fetching checksum file
+	response, err := http.Get(kubeNodeChecksumFileURL)
+	if err != nil {
+		return "", fmt.Errorf("could not fetch kube node binary checksum file: %v", err)
+	}
+	defer response.Body.Close()
+	// the checksum file content contains a single <sha> value
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: This comment would be better placed before `for scanner.Scan()` as that is where this assumption takes effect. You should also mention that you expect an EOF in the first line itself.

File: internal/test/wmcb/wmcb_test.go [hunk 7]
```diff
@@ -677,3 +677,42 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
+	var checksumFileContent string
+	// Fetching checksum file content from the GET Response
+	scanner := bufio.NewScanner(response.Body)
+	for scanner.Scan() {
+		checksumFileContent = scanner.Text()
+	}
+	if checksumFileContent == "" {
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: You should also check `scanner.Err()`

File: internal/test/wmcb/wmcb_test.go (hunk 6 above)
Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go]: Thinking a bit more on this, it would be safer for you to have a guard against the contents being in an unexpected format. Otherwise in the scenario when the checksum file contents have changed, the test would fail in an unexpected manner.

File: internal/test/wmcb/wmcb_test.go [hunk 8]
```diff
@@ -677,3 +677,42 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
 	sha512 := sha512AndFilename[0]
 	return sha512, nil
 }
+
+// initializeKubeNodePkgInfo populates the fileds of KubeNode pkg
+func (f *wmcbFramework) initializeKubeNodePkgInfo() error {
+	f.kubeNode = pkgInfo{}
+
+	f.kubeNode.url = kubeNodeBaseURL + framework.K8sVersion + "/kubernetes-node-windows-amd64.tar.gz"
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: Do you want to check if `K8sVersion` is not emtpy?

File: internal/test/wmcb/wmcb_test.go [hunk 9]
```diff
@@ -677,3 +677,42 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
+		return fmt.Errorf("unable to get SHA for Kubernetes node binary for Windows: %v", err)
+	}
+	f.kubeNode.sha = kubeNodeSHA
+
+	f.kubeNode.shaType = "sha512"
+	return nil
+}
+
+// getKubeNodeSHA returns the SHA512 value of the kube node package.
+func (f *wmcbFramework) getKubeNodeSHA() (string, error) {
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: Technically this could have been `kube package` struct method. No need of having it in framework right?

File: internal/test/wmcb/wmcb_test.go (hunk 6 above)
Comment (ravisantoshgudimetla) [reply internal/test/wmcb/wmcb_test.go]: It's also better to write an example of how contents would look like in the comments, so that while reviewing we can see if have more validations.

File: internal/test/framework/framework.go [hunk 10]
```diff
@@ -215,29 +220,34 @@ func (f *TestFramework) getOpenShiftOperatorClient(config *restclient.Config) er
...
 
 	// Convert kubernetes version to major.minor format. v1.17.0 -> 1.17
-	k8sVersion := strings.TrimLeft(versionInfo.GitVersion, "v")
-	k8sSemver := strings.Split(k8sVersion, ".")
+	k8sSemver := strings.Split(f.K8sVersion[1:], ".")
 	k8sMinorVersion, err := strconv.Atoi(k8sSemver[1])
 	if err != nil {
 		return fmt.Errorf("could not conver k8s minor version %s to int", err)
 	}
-
```
[the remark this answers is not in the document]
Comment (VaishnaviHire) [reply internal/test/framework/framework.go]: Should I rollback this change?

[the remark this answers is not in the document]
Comment (aravindhp) [reply internal/test/framework/framework.go]: Yes please.

File: internal/test/wmcb/wmcb_test.go [hunk 11]
```diff
@@ -677,3 +677,48 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
 	sha512 := sha512AndFilename[0]
 	return sha512, nil
 }
+
+// initializeKubeNodePkgInfo populates the fileds of KubeNode pkg
+func (f *wmcbFramework) initializeKubeNodePkgInfo() error {
+	f.kubeNode = pkgInfo{}
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: you can just initialize the pkgInfo{} and then set f.kubeNode to this pkgInfo{}

File: internal/test/wmcb/wmcb_test.go [hunk 12]
```diff
@@ -677,3 +677,48 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
+	if err != nil {
+		return fmt.Errorf("unable to get SHA for Kubernetes node binary for Windows: %v", err)
+	}
+	f.kubeNode.sha = kubeNodeSHA
+
+	f.kubeNode.shaType = "sha512"
+	return nil
+}
+
+// getKubeNodeSHA returns the SHA512 value of the kube node package.
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: Please add a TODO and a story for this.

Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go]: Please explain what `this` is.

Comment (ravisantoshgudimetla) [reply internal/test/wmcb/wmcb_test.go]: I had an offline discussion with @VaishnaviHire. We decided to use `pkgInfo` struct for now for getting the SHA. We should ideally use `pkgInfo{}` as interface and have a slice of `pkgInfo{}` interfaces instead of have 2 different structs for the same `pkgInfo`

Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go]: @ravisantoshgudimetla this again is super low priority work that will never get picked up. We either do this now or not at all. There is a limit to which we can keep generating TODOs and future work.

Comment (ravisantoshgudimetla) [reply internal/test/wmcb/wmcb_test.go]: Ok, I am not sure if this blocks @VaishnaviHire for closing this ticket before tomorrow. If we don't mind, I think we should have this as part of the PR.

Comment (VaishnaviHire) [reply internal/test/wmcb/wmcb_test.go]: I already added the TODOs. Should I include the refactoring as a part of this PR and implement the changes ?

Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go]: Please pair program.

File: internal/test/wmcb/wmcb_test.go [hunk 13]
```diff
@@ -198,9 +232,8 @@ func (vm *wmcbVM) initializeTestBootstrapperFiles() error {
 	if err != nil {
 		return fmt.Errorf("unable to create remote directory %s: %v", remoteDir, err)
 	}
-
 	// Download and extract the kube package on the VM
-	err = vm.remoteDownloadExtract(kubeNode, remoteDir+"kube.tar.gz", remoteDir)
+	err = vm.remoteDownloadExtract(framework.pkgs[0], remoteDir+"kube.tar.gz", remoteDir)
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: Can you loop over the slice instead of passing on the slice along with indexes. In fact do you think, if we have a `name` in pkgInfo{} struct to uniquely identify the item in the slice?

hint: You can also use a map.

File: internal/test/wmcb/wmcb_test.go [hunk 14]
```diff
@@ -87,11 +82,49 @@ type pkgInfo struct {
...
 	// TestFramework holds the instantiation of test suite being executed
 	*e2ef.TestFramework
-	// cniPlugins contains information about the CNI plugin package
-	cniPlugins pkgInfo
+	// contains array of packages to be downloaded
+	pkgs []pkgInfo
+}
+
+//kubeNode contains the information about  the kubernetes node package for Windows
+type KubeNode struct{
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: You don't need an explicit field for `pkg` in the struct. The struct can implement it.

File: internal/test/wmcb/wmcb_test.go [hunk 15]
```diff
@@ -87,11 +86,41 @@ type pkgInfo struct {
...
+	// get SHAvalue for the package
+	getSHAValue(url string) (string, error)
+}
+
 type wmcbFramework struct {
 	// TestFramework holds the instantiation of test suite being executed
 	*e2ef.TestFramework
-	// cniPlugins contains information about the CNI plugin package
-	cniPlugins pkgInfo
+	// contains array of packages to be downloaded
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: I think this comment needs to be corrected since you are using map instead of array now

File: internal/test/wmcb/wmcb_test.go [hunk 16]
```diff
@@ -87,11 +82,52 @@ type pkgInfo struct {
...
+	pkgs map[string]pkgInfo
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: nit: The maps's keys should not of types strings. Have a type definition for it.

File: internal/test/wmcb/wmcb_test.go [hunk 17]
```diff
@@ -635,28 +664,43 @@ func testWMCBCluster(t *testing.T) {
...
 	}
-	f.cniPlugins.sha = cniPluginsSHA
+	cniPluginsPkg.sha = cniPluginsSHA
 
-	f.cniPlugins.shaType = "sha512"
-	return nil
-}
+	cniPluginsPkg.shaType = "sha512"
+	c.pkgInfo = cniPluginsPkg.pkgInfo
+	return nil}
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: code formatting: closing bracket '}' can come on next line

File: internal/test/wmcb/wmcb_test.go [hunk 18]
```diff
@@ -87,11 +82,52 @@ type pkgInfo struct {
...
+	// intialize cni plugin package and add to the package map
+	if err :=  cniPluginsPackage.initializePkgInfo(); err != nil {
+		return fmt.Errorf("unable to initialize CNI Plugins package info: %v", err)
+	}
+	packages["cniPlugins"] = cniPluginsPackage.pkgInfo
+	// initialize kube Node binary package and add to the package map
+	if err := kubeNodePackage.initializePkgInfo(); err != nil{
+		return fmt.Errorf("unable to initialize kubernetes node package info for Windows: %v", err)
+	}
+	packages["kubeNode"] = kubeNodePackage.pkgInfo
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: After you create [packages](https://github.com/openshift/windows-machine-config-bootstrapper/pull/170/files#diff-631087f953268c655ad0d05d71b2f6e2R115) you can insert the key and make it tigther for validation.

File: internal/test/wmcb/wmcb_test.go [hunk 19]
```diff
@@ -87,11 +82,52 @@ type pkgInfo struct {
...
+	pkgs map[string]pkgInfo
+}
+
+// cniPlugins contains information about the CNI plugin package
+type CniPlugins struct {
+	pkgInfo
+}
+
+//kubeNode contains the information about  the kubernetes node package for Windows
+type KubeNode struct{
```
Comment (ravisantoshgudimetla) [inline internal/test/wmcb/wmcb_test.go]: Why are these public?

File: internal/test/wmcb/wmcb_test.go [hunk 20]
```diff
@@ -635,28 +664,43 @@ func testWMCBCluster(t *testing.T) {
...
+	sha512 := sha512AndFilename[0]
+	return sha512, nil
+}
+
+// getSHAFileContent returns the contents of the SHA file for the given url of the package.
+func (f *pkgInfo) getSHAFileContent(url string)(string, error){
+	pkgChecksumFileURL := url + ".sha512"
+	response, err := http.Get(pkgChecksumFileURL)
 	if err != nil {
 		return "", fmt.Errorf("could not fetch CNI Plugins checksum file: %s", err)
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: Do we fetch checksum of only CNI plugins here. Do we need to change error message?

File: internal/test/wmcb/wmcb_test.go [hunk 21]
```diff
@@ -87,11 +82,52 @@ type pkgInfo struct {
 	shaType string
 }
 
+// interface to populate pkInfo structs
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: pkgInfo structs

File: internal/test/wmcb/wmcb_test.go [hunk 22]
```diff
@@ -713,4 +720,37 @@ func (f *pkgInfo) getSHAFileContent(url string)(string, error){
 		checksumFileContent = scanner.Text()
 	}
 	return  checksumFileContent, nil
-}
\ No newline at end of file
+}
+
+// initializeKubeNodePkgInfo populates the fileds of KubeNode pkg
+func (f *KubeNode) initializePkgInfo() error {
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: Name of the function does not match with the one given in the comment.

File: internal/test/wmcb/wmcb_test.go [hunk 23]
```diff
@@ -713,4 +720,37 @@ func (f *pkgInfo) getSHAFileContent(url string)(string, error){
...
+}
+
+// getKubeNodeSHA returns the SHA512 value of the kube node package given the package url.
+func (f *KubeNode) getSHAValue(kubeNodeURL string) (string, error) {
+	checksumFileContent, err := f.getSHAFileContent(kubeNodeURL)
+	if err!= nil{
+		return "", err
+	}
+	// the checksum file content contains a single value in the "<sha>" format.
+	sha512 := strings.Split(checksumFileContent, "\n")
```
Comment (akhil-rane) [inline internal/test/wmcb/wmcb_test.go]: Why are you splitting if checksum file content has only <sha>. Are you trying to remove '\n' ? Can you use [`strings.TrimSpace`](https://golang.org/pkg/strings/#TrimSpace) instead?

Comment (mansikulkarni96) [reply internal/test/wmcb/wmcb_test.go]: I think she is trying to split the string on "\n"

File: internal/test/wmcb/wmcb_test.go [hunk 24]
```diff
@@ -635,30 +681,47 @@ func testWMCBCluster(t *testing.T) {
 	assert.Lenf(t, winNodes.Items, 1, "expected one node to have node label but found: %v", len(winNodes.Items))
 }
 
-// initializeCNIPluginsPkgInfo populates the fields of cniPlugins
-func (f *wmcbFramework) initializeCniPluginsPkgInfo() error {
-	f.cniPlugins = pkgInfo{}
+//intitializePkgInfo populates the fields for cniPlugins package
+func (c *cniPlugins) initializePkgInfo() error {
+	var cniPluginsPkg cniPlugins
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: I am confused about this. Why is this needed?

File: internal/test/wmcb/wmcb_test.go [hunk 25]
```diff
@@ -635,30 +681,47 @@ func testWMCBCluster(t *testing.T) {
...
 	if err != nil {
 		return fmt.Errorf("unable to get SHA for CNI Plugins: %v", err)
 	}
-	f.cniPlugins.sha = cniPluginsSHA
+	cniPluginsPkg.sha = cniPluginsSHA
 
-	f.cniPlugins.shaType = "sha512"
+	cniPluginsPkg.shaType = "sha512"
+	c.pkgInfo = cniPluginsPkg.pkgInfo
+	c.pkgName = cniPluginPkgName
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: This does not look elegant. I suggest having a `newCNIPlugins` function that does all of this initialization. `initialize()` can be removed and should not be part of the interface. 

I also don't understand why `getSHAValue()` and `getSHAFileContent()` is part of the interface as I don't see it being used outside of `initialize()`.

I guess the question I am asking is, why is the interface needed? It is not like you are exporting it from external use.

File: internal/test/wmcb/wmcb_test.go [hunk 26]
```diff
@@ -64,20 +66,21 @@ var (
...
-	// kubeNode contains the information about  the kubernetes node package for Windows
-	kubeNode = pkgInfo{
-		url:     "https://dl.k8s.io/v1.16.2/kubernetes-node-windows-amd64.tar.gz",
-		sha:     "a88e7a1c6f72ea6073dbb4ddfe2e7c8bd37c9a56d94a33823f531e303a9915e7a844ac5880097724e44dfa7f4a9659d14b79cc46e2067f6b13e6df3f3f1b0f64",
-		shaType: "sha512",
-	}
+	// kubeNodePkgName is the user-defined name of the required kube node package
+	kubeNodePkgName = pkgName("kubeNode")
+	// cniPluginPkgName is the user-defined name of the required cni plugins package package
+	cniPluginPkgName = pkgName("cniPlugins")
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: Nit: This looks like overkill but is most likely @ravisantoshgudimetla's suggestion ;-) 

Not a nit: This needs to go into the previous commit.

File: internal/test/wmcb/wmcb_test.go [hunk 27]
```diff
@@ -668,12 +731,39 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
-	// The checksum file content is in the format "<sha> <filename>". So to get SHA we need to extract only the <sha>
-	// from the file
-	sha512AndFilename := strings.Split(checksumFileContent, " ")
-	if len(sha512AndFilename) < 2 {
-		return "", fmt.Errorf("checksum file content is not in the format:'<sha> <filename>'")
+	return checksumFileContent, nil
+}
+
+// initializePkgInfo populates the fields of KubeNode pkg
+func (f *kubeNode) initializePkgInfo() error {
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: Receiver should be `k`

File: internal/test/wmcb/wmcb_test.go [hunk 28]
```diff
@@ -668,12 +731,39 @@ func (f *wmcbFramework) getLatestCniPluginsSHA() (string, error) {
...
+	kubeNodePkg.sha = kubeNodeSHA
+
+	kubeNodePkg.shaType = "sha512"
+	f.pkgInfo = kubeNodePkg.pkgInfo
+	f.pkgName = kubeNodePkgName
+	return nil
+}
+
+// getSHAValue returns the SHA512 value of the kube node package given the package url.
+func (f *kubeNode) getSHAValue(kubeNodeURL string) (string, error) {
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: Receiver should be `k`.

File: internal/test/wmcb/wmcb_test.go [hunk 29]
```diff
@@ -635,30 +681,47 @@ func testWMCBCluster(t *testing.T) {
...
+	sha512AndFilename := strings.Split(checksumFileContent, " ")
+	if len(sha512AndFilename) < 2 {
+		return "", fmt.Errorf("checksum file content is not in the format:'<sha> <filename>'")
+	}
+	sha512 := sha512AndFilename[0]
+	return sha512, nil
+}
+
+// getSHAFileContent returns the contents of the SHA file for the given url of the package.
+func (f *pkgInfo) getSHAFileContent(url string) (string, error) {
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go]: Receiver should be `p`

File: internal/test/framework/framework.go [hunk 30]
```diff
@@ -215,6 +220,16 @@ func (f *TestFramework) getOpenShiftOperatorClient(config *restclient.Config) er
 	return nil
 }
 
+// getK8sVersion gets the Kubernetes version in vx.x.x format
+func (f *TestFramework) getK8sVersion() error {
```
Comment (sebsoto) [inline internal/test/framework/framework.go]: There is overlapping functionality here with the getClusterVersion(). I would merge these two functions as they both have the purpose of getting the cluster version.

Comment (VaishnaviHire) [reply internal/test/framework/framework.go]: Done

File: internal/test/wmcb/pkginfo.go [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	// retrieve SHA value and set it to the sha field
+	if err := pkg.initSHA(); err != nil {
+		return nil, fmt.Errorf("error retrieving the SHA value for the package: %v", err)
+	}
+	return pkg, nil
+}
+
+// getSHAFileContent returns the contents of the SHA file for the given url of the package.
+func getSHAFileContent(url string) (string, error) {
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:127]: This should be associated with `pkgInfo` and when you do that, please check that `url != ""`.

File: internal/test/wmcb/pkginfo.go [hunk 32]
```diff
@@ -0,0 +1,110 @@
...
+	sha512 := strings.Split(checksumFileContent, " ")
+	if len(sha512) < 2 {
+		return fmt.Errorf("checksum file content is not in the format : '<sha>'")
+	}
+	c.sha = sha512[0]
+	return nil
+}
+
+// initializePkg sets all the fields for given pkg name and returns the
+func initializePkg(name pkgName, url string, version string) (PkgInfo, error) {
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:109]: The pattern followed in go is to call this `newPkgInfo()`.

File: internal/test/wmcb/wmcb_test.go [hunk 33]
```diff
@@ -87,11 +90,54 @@ type pkgInfo struct {
...
+	getSHAValue(url string) (string, error)
+}
+
 type wmcbFramework struct {
 	// TestFramework holds the instantiation of test suite being executed
 	*e2ef.TestFramework
-	// cniPlugins contains information about the CNI plugin package
-	cniPlugins pkgInfo
+	// contains map of packages to be downloaded
+	pkgs map[pkgName]pkgInfo
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go:82]: Doc has not been fixed

File: internal/test/wmcb/wmcb_test.go [hunk 34]
```diff
@@ -70,28 +68,36 @@ var (
...
+
+// initializePackages sets up all the required packages of type pkgInfo
+func (f *wmcbFramework) initializePackages() error {
+	var pkgs = make(map[pkgName]pkgInfo)
+	// create pkgInfo struct for cni plugins and populate it
+	cniPluginsPackage, err := initializePkg(CniPluginPkgName, cniPluginsBaseURL, framework.LatestCniPluginsVersion)
+	if err != nil {
+		return err
+	}
+	cniPkgInfo := cniPluginsPackage.(*cniPlugins).pkgInfo
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:93]: This variable is not needed. Moreover if you had used the interface in the `pkgs` map, you would not have needed to do this sort of casting. I suggest you go that route.

File: internal/test/wmcb/pkginfo.go [hunk 35]
```diff
@@ -73,6 +80,31 @@ func (c *cniPlugins) initSHA() error {
 	return nil
 }
 
+// initUrl retrieves kubenode binary url given the k8s version and sets it to url field of kubeNode
+func (k *kubeNode) initUrl(baseUrl string, version string) error {
+	if version == "" {
+		return fmt.Errorf("unable to retrieve latest version of cni plugins")
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:86]: Incorrect error message.

File: internal/test/wmcb/wmcb_test.go [hunk 36]
```diff
@@ -282,6 +289,7 @@ func (vm *wmcbVM) initializeTestConfigureCNIFiles(ovnHostSubnet string) error {
 
 	// Download and extract the CNI binaries on the Windows VM
 	err = vm.remoteDownloadExtract(framework.pkgs[CniPluginPkgName], remoteDir+path.Base(cniUrl.Path), winCNIDir)
+
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:292]: This new line is not required.

File: internal/test/wmcb/pkginfo.go [hunk 37]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// getPkgName returns the user defined name of the package
+func (p *pkgInfo) getPkgName() pkgName { return p.name }
+
+//getShaType returns the sha type of the package
+func (p *pkgInfo) getShaType() string { return p.shaType }
+
+//getPkgUrl returns the url of the package
+func (p *pkgInfo) getPkgUrl() string { return p.url }
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:154]: Please format for readability. `return ` should be on a new line. I know some codebases follow this, but I suggest we not go down that route.

File: internal/test/wmcb/pkginfo.go [hunk 38]
```diff
@@ -0,0 +1,144 @@
...
+	//getPkgName returns the user defined package name
+	getPkgName() pkgName
+	//getPkgUrl returns the url of the package
+	getPkgUrl() string
+	// getSHAFileContent returns checksum file content
+	getSHAFileContent() (string, error)
+	//getSHAValue returns SHA value for the package
+	getSHAValue() (string, error)
+	// getShaType returns SHA type (e.g sha256)
+	getShaType() string
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:39]: `getSHAType()` for consistency. Or use `Sha` everywhere.

File: internal/test/wmcb/pkginfo.go [hunk 39]
```diff
@@ -0,0 +1,144 @@
...
+func (p *pkgInfo) getSHAFileContent() (string, error) {
+	pkgChecksumFileURL := p.url + "." + p.shaType
+	response, err := http.Get(pkgChecksumFileURL)
+	if err != nil {
+		return "", fmt.Errorf("could not fetch content from checksum file: %s", err)
+	}
+	defer response.Body.Close()
+
+	var checksumFileContent string
+	// Fetching checksum file content from the GET Response
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:120]: Please address https://github.com/openshift/windows-machine-config-bootstrapper/pull/170#discussion_r398567428

File: internal/test/wmcb/pkginfo.go [hunk 40]
```diff
@@ -0,0 +1,144 @@
...
+	scanner := bufio.NewScanner(response.Body)
+	for scanner.Scan() {
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:122]: Please address https://github.com/openshift/windows-machine-config-bootstrapper/pull/170#discussion_r398244118

File: internal/test/wmcb/pkginfo.go [hunk 41]
```diff
@@ -0,0 +1,144 @@
...
+
+//hybridOverlay contains information about the hybrid overlay binary
+type hybridOverlay struct {
+	pkgInfo
+}
+
+// PkgInfo is an interface to populate pkgInfo structs
+type PkgInfo interface {
+	//getPkgName returns the user defined package name
+	getPkgName() pkgName
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:31]: `getName()`

File: internal/test/wmcb/pkginfo.go [hunk 42]
```diff
@@ -0,0 +1,144 @@
...
+	//getPkgUrl returns the url of the package
+	getPkgUrl() string
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:33]: `getUrl()`

File: internal/test/wmcb/pkginfo.go [hunk 43]
```diff
@@ -84,6 +92,21 @@ func newHybridOverlayPkg(name pkgName, shaType string) (PkgInfo, error) {
...
+		return nil, fmt.Errorf("k8s version is not specified")
+	}
+	if baseUrl == "" {
+		return nil, fmt.Errorf("base url for kube node binary is not specified")
+	}
+
+	return &kubeNode{pkgInfo{
+		name:    name,
+		shaType: shaType,
+		url:     baseUrl + version + "/kubernetes-node-windows-amd64.tar.gz"}}, nil
```
Comment (aravindhp) [inline internal/test/wmcb/pkginfo.go:107]: Same comment about readability as before and placement of braces.

File: internal/test/wmcb/pkginfo.go (hunk 39 above)
Comment (VaishnaviHire) [reply internal/test/wmcb/pkginfo.go:120]: Addressed the comment in getSHAValue() function, as content format is different for every package https://github.com/openshift/windows-machine-config-bootstrapper/pull/170/files#diff-6658479ed9056a08f91d4cf1b23fda4bR137  (Line137)

**Comment by mansikulkarni96** [comment]
lgtm, waiting for tests to pass.

**Comment by aravindhp** [comment]
/hold

To prevent bot retests while @VaishnaviHire figures out why the tests are failing.

**Comment by VaishnaviHire** [comment]
> /hold
> 
> To prevent bot retests while @VaishnaviHire figures out why the tests are failing.

This bug doesn`t block the PR from merging.Refer [WINC-349](https://issues.redhat.com/browse/WINC-349) to track the bug.

**Comment by aravindhp** [comment]
/hold

@VaishnaviHire I resolved the merge conflict using the web UI and it created a [merge](Merge branch 'master' into winc_235) commit which is not so desirable. Please fix locally and push to overwrite the changes. Sigh I should not have tried to help :-(

**Comment by ravisantoshgudimetla** [comment]
/hold cancel

Considering this was LGTM'ed a long time ago.
